# 🚀 Customer Genome Intelligence System
### Decode Customer Behavior • Predict Value • Measure Confidence


## Objective

In [1]:
import pandas as pd
import numpy as np

In [8]:
print("This notebook cleans and prepares transactional retail data for downstream modeling.")

This notebook cleans and prepares transactional retail data for downstream modeling.


## Dataset Loading

In [2]:
data_path = "../data/Online_Retail.xlsx"
df = pd.read_excel(data_path)
df.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom

## Initial Data Inspection

In [3]:
df.shape

(541909, 8)

In [4]:
df.columns

Index(['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'UnitPrice', 'CustomerID', 'Country'],
      dtype='object')

In [5]:
df.isnull().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

## Column Standardization

In [6]:
df = df.rename(columns={
    "CustomerID": "customer_id",
    "InvoiceDate": "invoice_date",
    "InvoiceNo": "invoice_no",
    "StockCode": "product_id",
    "Quantity": "quantity",
    "UnitPrice": "price"
})
df.head()

invoice_no product_id                          Description  quantity  \
0     536365     85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1     536365      71053                  WHITE METAL LANTERN         6   
2     536365     84406B       CREAM CUPID HEARTS COAT HANGER         8   
3     536365     84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4     536365     84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

         invoice_date  price  customer_id         Country  
0 2010-12-01 08:26:00   2.55      17850.0  United Kingdom  
1 2010-12-01 08:26:00   3.39      17850.0  United Kingdom  
2 2010-12-01 08:26:00   2.75      17850.0  United Kingdom  
3 2010-12-01 08:26:00   3.39      17850.0  United Kingdom  
4 2010-12-01 08:26:00   3.39      17850.0  United Kingdom

## Handling Missing and Invalid Records

In [7]:
df = df.dropna(subset=["customer_id"])
df = df[df["quantity"] > 0]
df = df[df["price"] > 0]
df.shape

(397884, 8)

## Datetime Conversion

In [9]:
df["invoice_date"] = pd.to_datetime(df["invoice_date"])
df.dtypes

invoice_no              object
product_id              object
Description             object
quantity                 int64
invoice_date    datetime64[ns]
price                  float64
customer_id            float64
Country                 object
dtype: object

## Feature Creation

In [10]:
df["transaction_value"] = df["quantity"] * df["price"]
df.head()

invoice_no product_id                          Description  quantity  \
0     536365     85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1     536365      71053                  WHITE METAL LANTERN         6   
2     536365     84406B       CREAM CUPID HEARTS COAT HANGER         8   
3     536365     84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4     536365     84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

         invoice_date  price  customer_id         Country  transaction_value  
0 2010-12-01 08:26:00   2.55      17850.0  United Kingdom              15.30  
1 2010-12-01 08:26:00   3.39      17850.0  United Kingdom              20.34  
2 2010-12-01 08:26:00   2.75      17850.0  United Kingdom              22.00  
3 2010-12-01 08:26:00   3.39      17850.0  United Kingdom              20.34  
4 2010-12-01 08:26:00   3.39      17850.0  United Kingdom              20.34

## Duplicate Record Removal

In [11]:
df = df.drop_duplicates()
df.shape

(392692, 9)

## Final Data Validation

In [13]:
df.describe(include="all")

invoice_no product_id                         Description  \
count     392692.0     392692                              392692   
unique     18532.0       3665                                3877   
top       576339.0     85123A  WHITE HANGING HEART T-LIGHT HOLDER   
freq         542.0       2023                                2016   
mean           NaN        NaN                                 NaN   
min            NaN        NaN                                 NaN   
25%            NaN        NaN                                 NaN   
50%            NaN        NaN                                 NaN   
75%            NaN        NaN                                 NaN   
max            NaN        NaN                                 NaN   
std            NaN        NaN                                 NaN   

             quantity                   invoice_date          price  \
count   392692.000000                         392692  392692.000000   
unique            NaN                            NaN            NaN   
top               NaN                            NaN            NaN   
freq              NaN                            NaN            NaN   
mean        13.119702  2011-07-10 19:13:07.771892480       3.125914   
min          1.000000            2010-12-01 08:26:00       0.001000   
25%          2.000000            2011-04-07 11:12:00       1.250000   
50%          6.000000            2011-07-31 12:02:00       1.950000   
75%         12.000000            2011-10-20 12:53:00       3.750000   
max      80995.000000            2011-12-09 12:50:00    8142.750000   
std        180.492832                            NaN      22.241836   

          customer_id         Country  transaction_value  
count   392692.000000          392692      392692.000000  
unique            NaN              37                NaN  
top               NaN  United Kingdom                NaN  
freq              NaN          349203                NaN  
mean     15287.843865             NaN          22.631500  
min      12346.000000             NaN           0.001000  
25%      13955.000000             NaN           4.950000  
50%      15150.000000             NaN          12.450000  
75%      16791.000000             NaN          19.800000  
max      18287.000000             NaN      168469.600000  
std       1713.539549             NaN         311.099224

| Aspect | Inference |
|------|-----------|
| Total Records | 392,692 retail transactions |
| Time Period | 01-Dec-2010 to 09-Dec-2011 |
| Countries Covered | 37 countries |
| Country Dominance | United Kingdom contributes ~89% transactions |
| Unique Invoices | 18,532 invoices |
| Unique Products | 3,665 products |
| Most Sold Product | WHITE HANGING HEART T-LIGHT HOLDER |
| Quantity (Median) | 6 units per transaction |
| Quantity Distribution | 75% orders ≤ 12 units |
| Quantity Outliers | Max quantity = 80,995 (bulk/anomaly) |
| Price (Median) | 1.95 per item |
| Price Range | 0.001 to 8,142.75 |
| Price Nature | Mostly low-priced items with rare high-price outliers |
| Transaction Value (Median) | 12.45 |
| Transaction Value Pattern | Right-skewed (few very large orders) |
| Max Transaction Value | 168,469.6 |
| Customer Behavior | Few customers contribute disproportionately high revenue |
| Data Quality Hint | Presence of extreme outliers → cleaning required |


## Processed Data Export

In [15]:
output_path = "../data/clean_transactions.csv"
df.to_csv(output_path, index=False)
output_path

'../data/clean_transactions.csv'

In [16]:
print("Cleaned dataset saved and ready for clustering, modeling, and dashboard inference.")

Cleaned dataset saved and ready for clustering, modeling, and dashboard inference.
